sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -687,10 +687,10 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
 
       case inSubqueryOrExistsSubquery =>
         plan match {
-          case _: Filter | _: SupportsSubquery | _: Join => // Ok
+          case _: SupportsSubquery | _: Join | _: UnaryNode => // Ok
```

Comment:
How to make sure that all UnaryNode can support IN/EXISTS subqueries (correlated and uncorrelated)? The test cases only covered Project, Aggregate, and Window. Maybe we could either only support these three nodes for now, or add more tests to cover other unary operators.

Also, would it work with Aggregate having correlated IN/EXISTS subqueries in the GROUP BY clause?

Reply:
The Exists/In subqueries in UnaryNodes are handled by inserting an existance join (or joins) bellow the UnaryNode and then the subquery is replaced with an attribute reference. This is a very generic pattern and should not depend on the actual UnaryNode. Yes, Filter is handled separately but that is a performance optimization, not a correctness issue (LeftSemi/LeftAnti are better than existance join).

But if you think that we should stay safe and only allow a subset of UnaryNodes, then I would be totally fine with supporting only the most common nodes - Project, Aggregate, Window and perhaps Sort.

I'll add some extra test cases for aggregate.

Reply:
I think it's safer to use a whitelist of nodes than to support all UnaryNodes, and it shouldn't be a common use case for EXISTS/IN subquery to be in Generate, Expand, Limit, etc. We can always add more nodes in the future if needed.